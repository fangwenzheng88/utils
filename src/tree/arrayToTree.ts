import { isFunction } from '../is'
import { TreeNode } from './types'

interface TreeConfig<T extends Record<string, any>> {
  key?: string
  parentKey?: string
  parentCondition?: string | null | undefined | number | ((data: T) => boolean)
}

const defaultTreeConfig: Required<TreeConfig<Record<string, any>>> = {
  key: 'key',
  parentKey: 'parentKey',
  parentCondition: -1
}

export function arrayToTree<T extends Record<string, any>>(array: T[], treeConfig?: TreeConfig<T>): TreeNode[] {
  const config = { ...defaultTreeConfig, ...treeConfig }
  const result: TreeNode[] = []
  const itemMap: Record<string, TreeNode> = {}

  for (const item of array) {
    const itemId = item[config.key]
    const parentItemId = item[config.parentKey]

    if (itemMap[itemId]) {
      itemMap[itemId] = {
        ...item,
        children: itemMap[itemId].children
      }
    } else {
      itemMap[itemId] = {
        ...item
      }
    }

    const treeItem = itemMap[itemId]

    const { parentCondition } = config
    if (isFunction(parentCondition) ? parentCondition(item) : parentCondition === parentItemId) {
      result.push(treeItem)
    } else {
      const children = itemMap[parentItemId]?.children ?? []
      children.push(treeItem)
      if (!itemMap[parentItemId]) {
        itemMap[parentItemId] = { children }
      } else {
        itemMap[parentItemId].children = children
      }
    }
  }

  return result
}
