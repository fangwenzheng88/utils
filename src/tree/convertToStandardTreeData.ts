export interface TreeDataStandardConfig<T = Record<string, any>, P = {}> {
  key?: string
  parentKey?: string
  title?: string
  children?: string
  /**
   *
   * @param node
   * @param level
   * @returns
   */
  extra?: (node: T, level: number) => P
}

export type TreeDataStandard<T = Record<string, any>, P = {}> = {
  key: string
  title: string
  parentKey?: string
  data: T
  children?: TreeDataStandard<T, P>[]
  level: number
} & P

/**
 * 将树形数据转换为标准的树形数据结构
 * @category tree/convertToStandardTreeData
 * @param treeData 原始树形数据
 * @param config 转换配置
 * @returns 标准的树形数据结构
 */
export function convertToStandardTreeData<T extends Record<string, any>, P extends Record<string, any> = {}>(
  treeData: readonly T[],
  config?: TreeDataStandardConfig<T, P>
): TreeDataStandard<T, P>[] {
  return convert(treeData, config)
}

function convert<T extends Record<string, any>, P extends Record<string, any> = {}>(
  treeData: readonly T[],
  config?: TreeDataStandardConfig<T, P>,
  parentKey: string | undefined = undefined,
  level: number = 0
): TreeDataStandard<T, P>[] {
  const defaultConfig: Required<TreeDataStandardConfig<T, P>> = {
    key: 'key',
    parentKey: 'parentKey',
    title: 'title',
    children: 'children',
    extra() {
      return {} as P
    }
  }
  const treeDataConfig: Required<TreeDataStandardConfig<T, P>> = { ...defaultConfig, ...config }

  return treeData.map((item) => {
    const key = item[treeDataConfig.key]
    const title = item[treeDataConfig.title]
    const data = item
    const initialChildren = item[treeDataConfig.children]
    let children: TreeDataStandard<T, P>[] | undefined
    if (initialChildren) {
      children = convert(initialChildren, treeDataConfig, key, level + 1)
    }
    const extraData = treeDataConfig.extra(item, level)

    return {
      key,
      parentKey,
      title,
      data,
      children,
      level,
      ...extraData
    }
  })
}
