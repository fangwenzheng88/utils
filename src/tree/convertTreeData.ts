import { isArray } from '../native-is'

/**
 * @category tree
 */
type ConvertedTreeData<T = Record<string, any>, P = {}> = {
  children?: ConvertedTreeData<T, P>[]
} & P

/**
 * 将树形数据转换为需要的格式
 * @category tree
 * @param treeData 原始树形数据
 * @returns 树形数据结构
 */
export function convertTreeData<T extends Record<string, any>, P extends Record<string, any>>(
  treeData: readonly T[],
  callback: (node: T, level: number) => P,
  childrenFieldName: keyof T = 'children',
  level: number = 0
): ConvertedTreeData<T, P>[] {
  if (!isArray(treeData)) {
    return []
  }

  return treeData.map((item) => {
    const initialChildren = item[childrenFieldName]
    let children: ConvertedTreeData<T, P>[] | undefined
    if (initialChildren) {
      children = convertTreeData(initialChildren, callback, childrenFieldName, level + 1)
    }

    const treeData = callback(item, level)

    return {
      ...treeData,
      children,
    }
  })
}
