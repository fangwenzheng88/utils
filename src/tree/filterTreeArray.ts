import type { TreeNode } from './types'

/**
 * 递归过滤树形结构数据中的节点，满足条件的节点和其所有字节点和其父节点都会被保留
 * 返回的数据是对treeDataArr数组中每一项的浅拷贝，不会修改原始treeDataArr中的值
 * @category tree/filterTreeArray
 * @param treeDataArr 树形结构数据数组
 * @param filterFn 过滤函数，用于判断节点是否符合条件，函数返回布尔值
 * @param childrenFieldName 子节点的字段名，默认为 'children'
 * @returns 过滤后的节点数组
 */
export function filterTreeArray<T extends TreeNode>(
  treeDataArr: readonly T[],
  filterFn: (node: T, parentNodes: T[]) => boolean,
  childrenFieldName: keyof T = 'children'
): T[] {
  return filterTreeDataRecursive(treeDataArr, filterFn, [], childrenFieldName)
}

/**
 * 内部递归函数，实际用于过滤树形结构数据的节点
 * @param treeDataArr 树形结构数据数组
 * @param filterFn 过滤函数，用于判断节点是否符合条件，函数返回布尔值
 * @param parentNodes 父节点数组
 * @param childrenFieldName 子节点的字段名，默认为 'children'
 * @returns 过滤后的节点数组
 */
function filterTreeDataRecursive<T extends TreeNode>(
  treeDataArr: readonly T[],
  filterFn: (node: T, parentNodes: T[]) => boolean,
  parentNodes: T[] = [],
  childrenFieldName: keyof T = 'children'
): T[] {
  const filteredNodes: T[] = []
  for (const node of treeDataArr) {
    const remainingNode = { ...node }
    const currentParentNodes = [...parentNodes, node]

    if (filterFn(node, currentParentNodes)) {
      const filteredNode: T = { ...remainingNode }
      filteredNodes.push(filteredNode)
    } else {
      const children = node[childrenFieldName]
      if (Array.isArray(children) && children.length > 0) {
        const filteredChildren = filterTreeDataRecursive(children, filterFn, currentParentNodes, childrenFieldName)

        if (filteredChildren.length > 0) {
          const remainingNodeWithFilteredChildren = {
            ...remainingNode,
            [childrenFieldName]: filteredChildren
          }
          filteredNodes.push(remainingNodeWithFilteredChildren)
        }
      }
    }
  }
  return filteredNodes
}
