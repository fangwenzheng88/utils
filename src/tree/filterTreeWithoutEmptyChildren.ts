import { isArray } from '../native-is'
import type { TreeNode } from './types'

/**
 * 过滤掉树状数据结构中没有子节点的节点
 *
 * 此函数接收一个树状数据结构的数组和一个可选的子节点字段名称作为参数
 * 它会遍历数组中的每个节点，如果节点有子节点（且子节点非空），则将该节点及其过滤后的子节点数组一起加入到结果数组中
 * 如果节点没有子节点，则直接将该节点加入到结果数组中
 *
 * @category tree
 * @param treeDataArr 树状数据结构的数组
 * @param childrenFieldName 子节点字段名称，默认为'children'
 * @returns 过滤后的树状数据结构数组
 *
 * @example
 * ```ts
 * filterTreeWithoutEmptyChildren(
 * [
 *   {
 *     key: '1',
 *     children: [
 *       {
 *         key: '1-1',
 *       },
 *       {
 *         key: '1-2',
 *         children: [],
 *       },
 *     ],
 *   },
 *   {
 *     key: '2',
 *     children: [],
 *   },
 * ]
 * )
 *
 * 输出：
 * [
 *   {
 *     key: '1',
 *     children: [
 *       {
 *         key: '1-1',
 *       },
 *     ],
 *   }
 * ]
 * ```
 */
export function filterTreeWithoutEmptyChildren<T extends TreeNode>(treeDataArr: T[], childrenFieldName: keyof T = 'children'): T[] {
  if (!isArray(treeDataArr)) {
    return []
  }
  const filteredNodes: T[] = []
  for (const node of treeDataArr) {
    const children = node[childrenFieldName]
    if (Array.isArray(children)) {
      const filteredChildren = filterTreeWithoutEmptyChildren(children, childrenFieldName)

      if (filteredChildren.length > 0) {
        const remainingNodeWithFilteredChildren = {
          ...node,
          [childrenFieldName]: filteredChildren,
        }
        filteredNodes.push(remainingNodeWithFilteredChildren)
      }
    } else {
      filteredNodes.push({ ...node })
    }
  }
  return filteredNodes
}
