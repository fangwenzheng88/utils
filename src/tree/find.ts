import type { TreeNode } from './types'

/**
 * 查找单个节点的函数
 * @category tree/find
 * @param nodes 节点数组
 * @param callback 回调函数，判断节点是否满足条件
 * @param childrenFieldName 子节点数组字段名，默认为 'children'
 * @returns 满足条件的节点，如果没有找到，则返回 null
 */
export function findTreeNode<T extends TreeNode>(nodes: T[], callback: (node: T) => boolean, childrenFieldName: keyof T = 'children'): T | undefined {
  for (const node of nodes) {
    if (callback(node)) {
      return node
    }

    const children = node[childrenFieldName] as T[]
    if (children) {
      const result = findTreeNode(children, callback)
      if (result) {
        return result
      }
    }
  }

  return undefined
}

/**
 * 查找多个节点的函数
 * @category tree/find
 * @param nodes 节点数组
 * @param callback 回调函数，判断节点是否满足条件
 * @param childrenFieldName 子节点数组字段名，默认为 'children'
 * @returns 满足条件的节点数组
 */
export function findTreeNodes<T extends TreeNode>(nodes: T[], callback: (node: T) => boolean, childrenFieldName: keyof T = 'children'): T[] {
  const results: T[] = []

  for (const node of nodes) {
    if (callback(node)) {
      results.push(node)
    }

    const children = node[childrenFieldName] as T[]
    if (children) {
      const childResults = findTreeNodes(children, callback)
      results.push(...childResults)
    }
  }

  return results
}

/**
 * 查找符合条件节点的父节点及祖先节点
 * @category tree/find
 * @param nodes 节点数组
 * @param callback 回调函数，判断节点是否满足条件，确保只有一个节点符合条件，如果多个节点符合条件，会返回多个节点的父节点及祖先节点
 * @param childrenFieldName 子节点数组字段名，默认为 'children'
 * @returns 满足条件的父节点数组
 */
export function findTreeParents<T extends TreeNode>(nodes: readonly T[], callback: (node: T) => boolean, childrenFieldName: keyof T = 'children'): T[] {
  const parents: T[] = []

  for (const node of nodes) {
    const children = node[childrenFieldName] as T[]
    if (children) {
      if (children.some(callback)) {
        parents.push(node)
      }

      const childParents = findTreeParents(children, callback, childrenFieldName)
      if (childParents.length > 0) {
        parents.push(node, ...childParents)
      }
    }
  }

  return parents
}

/**
 * 查找符合条件节点的父节点
 * @category tree/find
 * @param nodes 节点数组
 * @param callback 回调函数，判断节点是否满足条件，确保只有一个节点符合条件，如果多个节点符合条件，会返回多个节点的父节点
 * @param childrenFieldName 子节点数组字段名，默认为 'children'
 * @returns 满足条件的父节点数组
 */
export function findTreeParent<T extends TreeNode>(
  nodes: readonly T[],
  callback: (node: T) => boolean,
  childrenFieldName: keyof T = 'children'
): T | undefined {
  const parents: T[] = findTreeParents(nodes, callback, childrenFieldName)

  return parents.length > 0 ? parents[parents.length - 1] : undefined
}
