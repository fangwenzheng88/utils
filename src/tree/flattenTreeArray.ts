import type { TreeNode } from './types'

/**
 * 扁平化树结构数组
 * @category tree/flattenTreeArray
 * @param treeData 树结构数组
 * @param childrenFieldName 子节点字段名，默认为 'children'
 * @returns 扁平化后的树结构数组
 *
 * @example
 * ```typescript
 * interface MyTreeNode {
 *   id: number;
 *   name: string;
 *   children?: MyTreeNode[];
 * }
 *
 * const treeData: MyTreeNode[] = [
 *   {
 *     id: 1,
 *     name: 'Node 1',
 *     children: [
 *       {
 *         id: 2,
 *         name: 'Node 2',
 *       },
 *       {
 *         id: 3,
 *         name: 'Node 3',
 *         children: [
 *           {
 *             id: 4,
 *             name: 'Node 4',
 *           },
 *           {
 *             id: 5,
 *             name: 'Node 5',
 *           },
 *         ],
 *       },
 *     ],
 *   },
 *   {
 *     id: 6,
 *     name: 'Node 6',
 *   },
 * ];
 *
 * const flattenedData = flattenTreeArray(treeData);
 * ```
 */
export function flattenTreeArray<T extends TreeNode>(treeData: T[], childrenFieldName: keyof T = 'children'): T[] {
  const result: T[] = []
  for (const node of treeData) {
    result.push(node)
    const children = node[childrenFieldName]
    if (Array.isArray(children)) {
      result.push(...flattenTreeArray(children, childrenFieldName))
    }
  }
  return result
}
