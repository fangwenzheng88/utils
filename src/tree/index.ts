export * from './filterTreeArray'
export * from './sortTreeArray'
export * from './traverse'
export * from './find'
export * from './convertToStandardTreeData'
export * from './arrayToTree'
export * from './flattenTreeArray'

export type * from './types'
