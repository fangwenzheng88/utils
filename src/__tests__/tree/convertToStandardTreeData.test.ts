import { test, expect, describe, assertType } from 'vitest'
import { convertToStandardTreeData } from '../../tree'

interface TreeData {
  id: string
  name: string
  childrenNodes?: TreeData[]
}

const treeDataArr: TreeData[] = [
  {
    id: '1',
    name: 'Node 1',
    childrenNodes: [
      {
        id: '2',
        name: 'Node 1.1',
        childrenNodes: []
      },
      {
        id: '3',
        name: 'Node 1.2',
        childrenNodes: [
          { id: '4', name: 'Node 1.2.1' },
          { id: '5', name: 'Node 1.2.2' }
        ]
      }
    ]
  },
  {
    id: '6',
    name: 'Node 2',
    childrenNodes: []
  }
]

describe('convertToStandardTreeData', () => {
  test('转换非标准treeData为标准格式', () => {
    const result = convertToStandardTreeData(treeDataArr, {
      key: 'id',
      title: 'name',
      children: 'childrenNodes'
    })
    expect(result).toMatchSnapshot()
  })

  test('添加自定义字段', () => {
    const result = convertToStandardTreeData(treeDataArr, {
      key: 'id',
      title: 'name',
      children: 'childrenNodes',
      extra(node) {
        return { isLeaf: !Array.isArray(node.childrenNodes) }
      }
    })
    expect(result).toMatchSnapshot()
  })

  test('转换后类型测试', () => {
    const result = convertToStandardTreeData(treeDataArr, {
      key: 'id',
      title: 'name',
      children: 'childrenNodes',
      extra(node) {
        return { isLeaf: !Array.isArray(node.childrenNodes), number: 1 as const }
      }
    })

    assertType<TreeData['id']>(result[0].key)
    assertType<TreeData['name']>(result[0].title)
    assertType<boolean>(result[0].isLeaf)
    assertType<TreeData>(result[0].data)
    assertType<1>(result[0].number)
  })
})
