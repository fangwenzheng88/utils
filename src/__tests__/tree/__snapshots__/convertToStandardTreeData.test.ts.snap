// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convertToStandardTreeData > 添加自定义字段 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "data": {
          "childrenNodes": [],
          "id": "2",
          "name": "Node 1.1",
        },
        "isLeaf": false,
        "key": "2",
        "level": 1,
        "parentKey": "1",
        "title": "Node 1.1",
      },
      {
        "children": [
          {
            "children": undefined,
            "data": {
              "id": "4",
              "name": "Node 1.2.1",
            },
            "isLeaf": true,
            "key": "4",
            "level": 2,
            "parentKey": "3",
            "title": "Node 1.2.1",
          },
          {
            "children": undefined,
            "data": {
              "id": "5",
              "name": "Node 1.2.2",
            },
            "isLeaf": true,
            "key": "5",
            "level": 2,
            "parentKey": "3",
            "title": "Node 1.2.2",
          },
        ],
        "data": {
          "childrenNodes": [
            {
              "id": "4",
              "name": "Node 1.2.1",
            },
            {
              "id": "5",
              "name": "Node 1.2.2",
            },
          ],
          "id": "3",
          "name": "Node 1.2",
        },
        "isLeaf": false,
        "key": "3",
        "level": 1,
        "parentKey": "1",
        "title": "Node 1.2",
      },
    ],
    "data": {
      "childrenNodes": [
        {
          "childrenNodes": [],
          "id": "2",
          "name": "Node 1.1",
        },
        {
          "childrenNodes": [
            {
              "id": "4",
              "name": "Node 1.2.1",
            },
            {
              "id": "5",
              "name": "Node 1.2.2",
            },
          ],
          "id": "3",
          "name": "Node 1.2",
        },
      ],
      "id": "1",
      "name": "Node 1",
    },
    "isLeaf": false,
    "key": "1",
    "level": 0,
    "parentKey": undefined,
    "title": "Node 1",
  },
  {
    "children": [],
    "data": {
      "childrenNodes": [],
      "id": "6",
      "name": "Node 2",
    },
    "isLeaf": false,
    "key": "6",
    "level": 0,
    "parentKey": undefined,
    "title": "Node 2",
  },
]
`;

exports[`convertToStandardTreeData > 转换非标准treeData为标准格式 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "data": {
          "childrenNodes": [],
          "id": "2",
          "name": "Node 1.1",
        },
        "key": "2",
        "level": 1,
        "parentKey": "1",
        "title": "Node 1.1",
      },
      {
        "children": [
          {
            "children": undefined,
            "data": {
              "id": "4",
              "name": "Node 1.2.1",
            },
            "key": "4",
            "level": 2,
            "parentKey": "3",
            "title": "Node 1.2.1",
          },
          {
            "children": undefined,
            "data": {
              "id": "5",
              "name": "Node 1.2.2",
            },
            "key": "5",
            "level": 2,
            "parentKey": "3",
            "title": "Node 1.2.2",
          },
        ],
        "data": {
          "childrenNodes": [
            {
              "id": "4",
              "name": "Node 1.2.1",
            },
            {
              "id": "5",
              "name": "Node 1.2.2",
            },
          ],
          "id": "3",
          "name": "Node 1.2",
        },
        "key": "3",
        "level": 1,
        "parentKey": "1",
        "title": "Node 1.2",
      },
    ],
    "data": {
      "childrenNodes": [
        {
          "childrenNodes": [],
          "id": "2",
          "name": "Node 1.1",
        },
        {
          "childrenNodes": [
            {
              "id": "4",
              "name": "Node 1.2.1",
            },
            {
              "id": "5",
              "name": "Node 1.2.2",
            },
          ],
          "id": "3",
          "name": "Node 1.2",
        },
      ],
      "id": "1",
      "name": "Node 1",
    },
    "key": "1",
    "level": 0,
    "parentKey": undefined,
    "title": "Node 1",
  },
  {
    "children": [],
    "data": {
      "childrenNodes": [],
      "id": "6",
      "name": "Node 2",
    },
    "key": "6",
    "level": 0,
    "parentKey": undefined,
    "title": "Node 2",
  },
]
`;

exports[`convertToStandardTreeData > 转换非标准treeData为标准格式，并添加自定义字段 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "data": {
          "childrenNodes": [],
          "id": "2",
          "name": "Node 1.1",
        },
        "isLeaf": false,
        "key": "2",
        "level": 1,
        "parentKey": "1",
        "title": "Node 1.1",
      },
      {
        "children": [
          {
            "children": undefined,
            "data": {
              "id": "4",
              "name": "Node 1.2.1",
            },
            "isLeaf": true,
            "key": "4",
            "level": 2,
            "parentKey": "3",
            "title": "Node 1.2.1",
          },
          {
            "children": undefined,
            "data": {
              "id": "5",
              "name": "Node 1.2.2",
            },
            "isLeaf": true,
            "key": "5",
            "level": 2,
            "parentKey": "3",
            "title": "Node 1.2.2",
          },
        ],
        "data": {
          "childrenNodes": [
            {
              "id": "4",
              "name": "Node 1.2.1",
            },
            {
              "id": "5",
              "name": "Node 1.2.2",
            },
          ],
          "id": "3",
          "name": "Node 1.2",
        },
        "isLeaf": false,
        "key": "3",
        "level": 1,
        "parentKey": "1",
        "title": "Node 1.2",
      },
    ],
    "data": {
      "childrenNodes": [
        {
          "childrenNodes": [],
          "id": "2",
          "name": "Node 1.1",
        },
        {
          "childrenNodes": [
            {
              "id": "4",
              "name": "Node 1.2.1",
            },
            {
              "id": "5",
              "name": "Node 1.2.2",
            },
          ],
          "id": "3",
          "name": "Node 1.2",
        },
      ],
      "id": "1",
      "name": "Node 1",
    },
    "isLeaf": false,
    "key": "1",
    "level": 0,
    "parentKey": undefined,
    "title": "Node 1",
  },
  {
    "children": [],
    "data": {
      "childrenNodes": [],
      "id": "6",
      "name": "Node 2",
    },
    "isLeaf": false,
    "key": "6",
    "level": 0,
    "parentKey": undefined,
    "title": "Node 2",
  },
]
`;
